Cover: Video - Add position controls

ItsJonQ opened the pull request:
![focal-demo](https://user-images.githubusercontent.com/2322354/82595085-05daef80-9b73-11ea-9a2c-d2560084b371.gif)

This update adds positioning controls for video backgrounds within the **Cover block**. Previous to this image, the positioning controls (powered by the `FocalPointPicker` component) were available only to image backgrounds.

## ✨ New Features

##### Click to reposition

The position values can be updated by clicking anywhere in the preview area.

##### Keyboard arrow key support

![Screen Capture on 2020-05-21 at 14-53-25](https://user-images.githubusercontent.com/2322354/82595317-65d19600-9b73-11ea-8ed2-a5210db75ed6.gif)

Up, Down, Left, Right keyboard keys will adjust the position values.

##### Hold Shift to jump by 10

Holding <kbd>Shift</kbd> will jump the values, rounded to the nearest `10`. This works for boht dragging and keyboard arrow inputs.

##### New InputControls

The `top` and `left` values use a new `NumberControl` component. This component includes many features, such as dragging to update, jumping values by holding <kbd>Shift</kbd>, and automatic number rounding.

##### Grid

A 9x9 grid appears anytime a value is updated by **any** interaction (drag, click, etc...)

## 🛠 Fixes

* Enable `<FocalPointPicker />` to reset values based on undo actions in the editor.


## How has this been tested?

This has been tested locally in Storybook and local Gutenberg.

To test...

* Pull down the branch
* Run `npm install`
* Run `npm run dev` to fire up local Gutenberg
* Add a Cover block
* Add a video background (Here's a [sample video to download](https://www.w3schools.com/html/mov_bbb.mp4))
* Adjust the position
* Ensure it works in the editor and on the front-end of the site


## Types of changes

* Add video preview support to `FocalPointPicker`
* Substantial refactors to `FocalPointPicker`
* Adds `Flex` component to `@wordpress/components` as a layout-based primitive
* Update Cover video position rendering in `edit.js` and `save.js`

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Resolves: https://github.com/WordPress/gutenberg/issues/22509

Commit:
Add initial support for video in FocalPointPicker

Commit:
Add FocalPointPicker stories + update README.

Commit:
Improve FocalPointPicker and add Grid

Commit:
Update docs manifest

Commit:
Add click + keyboard arrow movements to update position

Commit:
Show grid on drag. Improve click/drag interaction.

Commit:
Apply Cover background position on save.js

Commit:
Refactor FocalPointPicker class component. Adjust styles.

Commit:
Refactor FocalPointPicker Media and add tests

Commit:
Refactor FocalPointPicker updateValue parsing.

Improve grid CSS rendering.

Commit:
Refactor percentage value updates from FocalPointPicker and sub components

ItsJonQ commented on packages/components/src/focal-point-picker/styles/focal-point-style.js:
```diff
@@ -0,0 +1,48 @@
+ * Internal dependencies
+ */
+import { color } from '../../utils/style-mixins';
+
```
Migrated and improved styles from the previous `styles.scss` file.

ItsJonQ commented:
@jasmussen Provided a design suggestion in the issue. Will attempt it now 🙌 

![](https://user-images.githubusercontent.com/1204802/82664585-49486300-9c32-11ea-83b0-5bcf93a2d36c.png)

Commit:
Add side label alignment for InputControl. Update controls in FocalPointPicker

Commit:
Add shift + drag to jump by 10. Improve stability of calculations on load

Commit:
Update fixture for Cover

ItsJonQ commented:
@jasmussen We have "side" label alignments for Inputs :D

<img width="254" alt="Screen Shot 2020-05-22 at 1 35 08 PM" src="https://user-images.githubusercontent.com/2322354/82694336-194c9000-9c31-11ea-8749-5bde37208fff.png">

Note! This is done at the `InputControl` component level. That means that we can more easily side-align labels in the future in a consistent manner (if we choose to)

ItsJonQ commented:
@jasmussen Based on your mockup, do you think we should remove the extra padding around the media (when possible)?

<img width="475" alt="Screen Shot 2020-05-22 at 1 38 00 PM" src="https://user-images.githubusercontent.com/2322354/82694592-82340800-9c31-11ea-88bc-716def0d91d3.png">

Doing so would change the height of the preview. We could set min/max values to ensure it doesn't get too wonky.

It comes down to... do we want the preview to have a consistent height (currently `200px`) or not 😊 

Would love your thoughts!

Commit:
Fix file path for media import in test

Commit:
Update Flex component from master

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

Commit:
Add .ogv for video file

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

Commit:
Format style files

jasmussen commented:
> @jasmussen Based on your mockup, do you think we should remove the extra padding around the media (when possible)?

OMG @ItsJonQ I'm so sorry I missed this ping! I don't know hos that got lost. Sorry about that.

Yes, I think we should remove both the border, and the inner padding — I can't quite figure out what that padding is for, if it serves a purpose then I can add a new design.

Border too, yes, I know that we might end up with a situation where a white background bleeds in. If it turns out to be a problem, we can fo a subtler shadow border that should only be visible in those cases, such as `box-shadow: inset 0 0 0 $border-width rgba($black, 0.4)`; or something like that.

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

Commit:
Fix FocalPointPicker to handle reset value cases (e.g. UNDO)

Commit:
Improve rendering styles of FocalPointPicker media and picker

ItsJonQ commented:
@jasmussen No worries! Thanks for the followup :)

I've updated the UI to remove the excess padding, and added a subtle border (via box-shadow):

<img width="259" alt="Screen Shot 2020-06-08 at 3 48 08 PM" src="https://user-images.githubusercontent.com/2322354/84073911-ac125c00-a99f-11ea-92eb-2861aae794cf.png">

Lemme know how it feels!

P.S. Excuse ma spooky 👻 cover image, haha

jasmussen commented:
This is way better.

All this needs is a code review 👍 👍

ItsJonQ commented:
@jasmussen Amazing! Thank you <3

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

Commit:
Fix import path for roundClamp

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

youknowriad commented on packages/components/src/focal-point-picker/styles/focal-point-picker-style.js:
```diff
@@ -0,0 +1,101 @@
+
+export const UnitControl = styled( BaseUnitControl )`
+	width: 100px;
+`;
```
So for my own knowledge, this means the sub-styles are passed as a `style` prop right? Are there any other magic-tricks done here? 

Also, this seems to force us to add  a `style` prop for most of our components. How do you see React native support fitting here. One principle we had for some time in `@wordpress/components` is that the public API for our components should  be as DOM agnostic as possible.

ItsJonQ replied:
`styled()` generated CSS are passed as classNames ✌️ .

```
styled.div`...`

// or

styled(SomeComponent)`...`
```

Creates...

```
<div className="css-12xjx1" />
```

With the generated CSS injected (and managed) as:

```
<style>
    ...
    .css-12xjx1 {...}
    ...
</style>
```

Hope this helps!

youknowriad replied:
Oh cool! How does this translate in react native?

ItsJonQ replied:
I believe it generates a React Native `Stylesheet` under the hood:

https://emotion.sh/docs/@emotion/native

https://github.com/emotion-js/emotion/blob/master/packages/native/src/styled.js#L7

I've used `styled` from `@emotion/native` in a pure React Native project. It worked great! However, I've personally not worked on project that required the exact same (styled) component to be compatible for both Web and Native.

That would be an interesting challenge!

youknowriad replied:
Would be cool to see if these new components work well for the mobile team.

youknowriad commented on packages/components/src/focal-point-picker/focal-point.js:
```diff
@@ -0,0 +1,49 @@
+export default function FocalPoint( {
+	coordinates = { left: '50%', top: '50%' },
+	isDragging = false,
+	...props
```
I noticed you do that often (spread extra props) and I'm having trouble figuring out why or whether I consider this a good or bad practice. Are these needed? I know I personally try to avoid them if possible as they can lead to react warnings and can be hard to follow.

ItsJonQ replied:
For component libraries, prop spreading is common and often the default. The reason is because the consumer should be able to use HTML approved properties (like `aria-*` or `data-*`) just like they would on any HTML element. Not only that, but `on*` event callbacks should be available as well.

A (mostly) reliable solution for handling React warnings is that inner components often have prop filtering built in. In the case of components created with `@emotion/styled`, they filter out non React/HTML props before passing it down to elements like `<div />`, `<input />`, etc...

youknowriad replied:
> For component libraries, prop spreading is common and often the default. The reason is because the consumer should be able to use HTML approved properties (like aria-* or data-*) just like they would on any HTML element. Not only that, but on* event callbacks should be available as well.

When I use `FocalPointPicker` as a user, I don't know its internals whether it's a div, a button, or just a string... which makes me think, I don't know what HTML  prop are approved. This also makes it harder to be compatible with mobile. A block doing `< FocalPointPicker some-dom-specific-prop="something" />` won't translate properly to mobile. 

It makes me wonder if we should try to move away from these iteratively.

youknowriad commented on packages/components/src/focal-point-picker/grid.js:
```diff
@@ -0,0 +1,79 @@
+} from './styles/focal-point-picker-style';
+
+export default function FocalPointPickerGrid( {
+	bounds = {},
```
Do you think bounds is better than height/width? I'm mainly thinking in terms of consistency or should it be `dimensions` like the prop for `FocalPointPicker`?

youknowriad commented on packages/components/src/focal-point-picker/grid.js:
```diff
@@ -0,0 +1,79 @@
+
+export default function FocalPointPickerGrid( {
+	bounds = {},
+	percentages = {
```
Seems like the only thing we need here is a new instance of the value when it changed. Do you think we should just name this `value` and avoid any computation/destructuring? As written it gives the impression that we do need each value separately but by looking at the code it's not really the case, we just need a new instance.

ItsJonQ replied:
Works for me

youknowriad commented on packages/components/src/focal-point-picker/grid.js:
```diff
@@ -0,0 +1,79 @@
+		}
+
+		return () => clearTimer();
+	}, [ value ] );
```
Seems like this could be simplied to just:

```
useEffect( () => {
 	setIsActive( true );
 	const timeout = setTimeout( () => {
 		setIsActive( false );
 	}, 600 );

 	return () => window.clearTimeout( timeout );
 }, [ value ] );
```

and avoid all refs. Am I missing anything?

ItsJonQ replied:
That works. I initially created refs so that the animation would not flash on first render. I was able to combine your solution with the `useUpdateEffect` hook from another PR:

https://github.com/WordPress/gutenberg/blob/8a35791d60f8a69e2d155990e9c7e564248a4713/packages/components/src/utils/use-update-effect.js

youknowriad commented on packages/components/src/focal-point-picker/index.js:
```diff
@@ -9,62 +10,86 @@ import classnames from 'classnames';
-		super( props );
-		this.onMouseMove = this.onMouseMove.bind( this );
+	constructor( props, context ) {
+		super( props, context );
```
in general we used to use this pattern `super( ...arguments )` but it's the same thing.

ItsJonQ replied:
That makes sense to me 👍

youknowriad commented on packages/components/src/input-control/index.js:
```diff
@@ -37,6 +37,7 @@ export function InputControl(
 		isPressEnterToChange = false,
 		isFloatingLabel = false,
 		label,
+		labelPosition = 'top',
```
Do you see this as something available in all *Control components?

ItsJonQ replied:
Unsure yet! However, I know there are certain UIs where a top vs bottom label position would be beneficial. ColorPicker is a good example. Also BoxControl. I guess we'll see :)

youknowriad commented on packages/e2e-tests/fixtures/blocks/core__cover__deprecated-4.serialized.html:
```diff
@@ -1,5 +1,5 @@
 <!-- wp:cover {"url":"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAACklEQVR4nGMAAQAABQABDQottAAAAABJRU5ErkJggg==","id":35,"focalPoint":{"x":"0.07","y":"0.07"}} -->
-<div class="wp-block-cover has-background-dim" style="background-image:url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAACklEQVR4nGMAAQAABQABDQottAAAAABJRU5ErkJggg==);background-position:7% 7%"><div class="wp-block-cover__inner-container"><!-- wp:paragraph {"align":"center","placeholder":"Write title…","fontSize":"large"} -->
+<div class="wp-block-cover has-background-dim" style="background-image:url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAACklEQVR4nGMAAQAABQABDQottAAAAABJRU5ErkJggg==);background-position:7.000000000000001% 7.000000000000001%"><div class="wp-block-cover__inner-container"><!-- wp:paragraph {"align":"center","placeholder":"Write title…","fontSize":"large"} -->
```
Seems like this change is an indication that something is wrong with the "rounding" of the focal point value.

ItsJonQ replied:
This is really odd... looking into this now!

ItsJonQ replied:
Fixed! I needed to wrap the values with `Math.round` in `block/save.js`

youknowriad reviewed (commented):
That component is still a monster :) but It's a nicer monster with these focused subcomponents.

Commit:
Merge branch 'master' into try/cover-video-background-focal-points

ItsJonQ commented on packages/components/src/focal-point-picker/controls.js:
```diff
@@ -0,0 +1,70 @@
+function UnitControl( props ) {
+	return (
+		<BaseUnitControl
+			className="focal-point-picker__controls-position-unit-control"
```
`className` is not required for styling ✌️. However, I kept them to maintain current style workflows + expectations within the project.

ItsJonQ replied:
Oh? That works for me 👍 . It would encourage a more component (API) driven approach for adjustments.

I know that ad-hoc className targeting has proved to be a bit problematic in the block/editor part of the stylesheets. I also understand that those were the most practical solutions at the time.

ItsJonQ commented on packages/block-library/src/cover/save.js:
```diff
@@ -104,6 +115,7 @@ export default function save( { attributes } ) {
 					loop
 					playsInline
 					src={ url }
+					style={ { objectPosition: positionValue } }
```
Ah, good catch. I'll make that update 👍

Commit:
Ensure Cover save does not generated empty styles if there are no video bg positions

Commit:
Remove unused focal-point-picker style

Commit:
Adjusted super call in FocalPointPicker constructor

Commit:
Adjust useRevealAnimation hook with useUpdateEffect for debouncing

ItsJonQ commented on packages/block-library/src/cover/edit.js:
```diff
@@ -327,28 +331,36 @@ function CoverEdit( {
-							) }
-						{ VIDEO_BACKGROUND_TYPE === backgroundType && (
-							<video autoPlay muted loop src={ url } />
+						{ isImageBackground && ! hasParallax && (
```
Oh, that works 👍 .  I'll created another `const` above. Something like `const showFocalPointPicker` so that it's easier to follow.

Commit:
Simplifiy showFocalPointPicker logic in cover/edit

Commit:
Fix position rounding for cover/save.js

youknowriad reviewed (approved):
This is looking very good 👍

ItsJonQ commented:
@youknowriad Thank you for reviewing this @youknowriad + @jasmussen  ! I'll merge it up 🙌

youknowriad commented:
Looks like this PR might have introduced a JavaScript unit test error on master 😬

ItsJonQ commented:
@youknowriad Uh oh!! Will look into that now

ItsJonQ commented:
@youknowriad PR incoming with fix~!
